### PR TITLE
add NCE to MP-CNN

### PR DESCRIPTION
**Pull request by Victor0118**
@tuzhucheng  Could you review it?

Thanks!

### CODE REVIEW

File: nce/NCE_MP_Pytorch/README.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
+# MP-CNN PyTorch Implementation
```
Comment (tuzhucheng) [inline nce/NCE_MP_Pytorch/README.md]: This basically copies what I had for MP-CNN, but this is NCE on top of MP-CNN. You should update everything in this README and indicate clearly if some measurement is MP-CNN only or MP-CNN with NCE.

**Comment by Victor0118** [comment]
Whoever are familiar or want to be familiar with MP-CNN/NCE are welcomed!

Thanks

File: nce/NCE_MP_Pytorch/trainers/sick_trainer.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
+import time
+
+import torch
+import torch.nn.functional as F
+from torch.optim.lr_scheduler import ReduceLROnPlateau
+
+from nce.NCE_MP_Pytorch.trainers.trainer import Trainer
```
Comment (tuzhucheng) [inline nce/NCE_MP_Pytorch/trainers/sick_trainer.py]: python module name should be all lower case

**Comment by Victor0118** [comment]
@tuzhucheng  duplicate files removed. Please check again.

File: mp_cnn/evaluation.py [hunk 3]
```diff
@@ -2,7 +2,8 @@
 from mp_cnn.evaluators.msrvid_evaluator import MSRVIDEvaluator
 from mp_cnn.evaluators.trecqa_evaluator import TRECQAEvaluator
 from mp_cnn.evaluators.wikiqa_evaluator import WikiQAEvaluator
-
+from nce.NCE_MP_Pytorch.evaluators.trecqa_evaluator import TRECQAEvaluator_NCE
```
Comment (tuzhucheng) [inline mp_cnn/evaluation.py]: There's some repetition in the module path.

`from nce.mp_cnn.evaluators.trecqa_evaluator` seems better to me, module path should not have capital letters and we already know it is PyTorch so that is redundant.

Please do this for all imports.

File: mp_cnn/trainers/trainer.py [hunk 4]
```diff
@@ -10,10 +10,15 @@ def __init__(self, model, train_loader, trainer_config, train_evaluator, test_ev
         self.train_loader = train_loader
         self.batch_size = trainer_config['batch_size']
         self.log_interval = trainer_config['log_interval']
+        self.dev_log_interval = trainer_config['dev_log_interval']
```
Comment (tuzhucheng) [inline mp_cnn/trainers/trainer.py]: why do we need a different log_interval for each dataset, can the `self.log_interval` be used from the line above?

File: nce/NCE_MP_Pytorch/.gitignore [hunk 5]
```diff
@@ -0,0 +1,101 @@
+# Byte-compiled / optimized / DLL files
```
[the remark this answers is not in the document]
Comment (tuzhucheng) [reply nce/NCE_MP_Pytorch/.gitignore]: Please address this comment. Move them to root .gitignore and do not repeat lines are already present in root .gitignore.

File: nce/NCE_MP_Pytorch/README.md (hunk 1 above)
Comment (tuzhucheng) [reply nce/NCE_MP_Pytorch/README.md]: Please address this comment and do not copy the MP-CNN README. I would add a new row for each dataset with NCE in the mp_cnn README.

File: nce/NCE_MP_Pytorch/evaluators/qa_evaluator.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import numpy as np
```
Comment (tuzhucheng) [inline nce/NCE_MP_Pytorch/evaluators/qa_evaluator.py]: What's the difference between this evaluator and the one I have for MP-CNN? They are both trying to calculate map and mrr right?

File: nce/NCE_MP_Pytorch/model.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        neg = self.negModel(input[1].sentence_1, input[1].sentence_2, input[1].ext_feats)
+        pos = self.dropout(pos)
+        neg = self.dropout(neg)
+        pos = self.linearLayer(pos)
+        neg = self.linearLayer(neg)
+        combine = torch.cat([pos, neg], 1)
+        return combine
+
+
+class MPCNN(nn.Module):
```
Comment (tuzhucheng) [inline nce/NCE_MP_Pytorch/model.py]: This code is still repeated. Please import it.

File: nce/NCE_MP_Pytorch/train_script.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
+import argparse
```
Comment (tuzhucheng) [inline nce/NCE_MP_Pytorch/train_script.py]: Document how to use this.

File: nce/NCE_MP_Pytorch/trainers/qa_trainer.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+            max_len_a = 0
+
+            batch_near_list = []
+            batch_qid = []
+            batch_aid = []
+
+            for i in range(batch.batch_size):
+                label_i = batch.label[i].cpu().data.numpy()[0]
+                question_i = batch.sentence_1[i]
+                # question_i = question_i[question_i!=1] # remove padding 1 <pad>
```
Comment (tuzhucheng) [inline nce/NCE_MP_Pytorch/trainers/qa_trainer.py]: Please delete all commented out code.

File: nce/NCE_MP_Pytorch/trainers/trecqa_trainer.py [hunk 10]
```diff
@@ -0,0 +1,7 @@
+from nce.NCE_MP_Pytorch.trainers.qa_trainer import QATrainer
+
+
+class TRECQATrainer_NCE(QATrainer):
```
Comment (tuzhucheng) [inline nce/NCE_MP_Pytorch/trainers/trecqa_trainer.py]: Please read Python convention on how to name classes here: https://www.python.org/dev/peps/pep-0008/#class-names and update names to adhere to this convention in all places in the PR.

Please do not use snake case (no underscore), change this to `TRECQATrainerNCE`.

File: nce/NCE_MP_Pytorch/README.md (hunk 1 above)
Comment (tuzhucheng) [reply nce/NCE_MP_Pytorch/README.md]: Or you can make a link to the MP-CNN README and do not repeat instructions that are already there. In this README just have the tables showing performance with and without NCE.

File: mp_cnn/evaluation.py (hunk 3 above)
Comment (Victor0118) [reply mp_cnn/evaluation.py]: NCE_MP_Pytorch is a folder. I need to change them to the smaller case. So it means folder name is supposed to be the smaller case, right?

File: mp_cnn/trainers/trainer.py (hunk 4 above)
Comment (Victor0118) [reply mp_cnn/trainers/trainer.py]: `self.log_interval` is the interval for each regular log and `dev_log_interval` is the interval between two validations.
In your MP-CNN code, you do the evaluation every epoch. I think it could be adjusted since sometimes an epoch is too long such as for TrecQA. Then I set a small dev_log_interval so that I get more results. This is also done in SM-CNN code.

File: nce/NCE_MP_Pytorch/evaluators/qa_evaluator.py (hunk 6 above)
Comment (Victor0118) [reply nce/NCE_MP_Pytorch/evaluators/qa_evaluator.py]: There is a minor difference in the model prediction. Moreover, no loss number is output during this step in NCE version.

File: mp_cnn/evaluation.py (hunk 3 above)
Comment (tuzhucheng) [reply mp_cnn/evaluation.py]: yep change folder name to all lower case, and do not repeat things already in the path like nce if this is already in nce parent folder and do not add pytorch anywhere since it's obvious.

File: nce/NCE_MP_Pytorch/model.py (hunk 7 above)
Comment (Victor0118) [reply nce/NCE_MP_Pytorch/model.py]: Not really... I delete the final three layers (dropout, linear and LogSoftmax) and forward() outputs a feature matrix with size (N, hidden_layer_units). Maybe I need to add a NCE flag here to select whether I will add these layers.

File: nce/NCE_MP_Pytorch/evaluators/qa_evaluator.py (hunk 6 above)
Comment (tuzhucheng) [reply nce/NCE_MP_Pytorch/evaluators/qa_evaluator.py]: I would suggest making a general evaluator that both your code and my code can use. If your's is more general then I can use your's. Also it doesn't hurt to return loss if you just don't use it?

File: nce/NCE_MP_Pytorch/model.py (hunk 7 above)
Comment (tuzhucheng) [reply nce/NCE_MP_Pytorch/model.py]: What you can do in this case is you can subclass MP-CNN and just override whatever you need to override. The goal is as little duplicate code as possible.

File: nce/NCE_MP_Pytorch/evaluators/qa_evaluator.py (hunk 6 above)
Comment (Victor0118) [reply nce/NCE_MP_Pytorch/evaluators/qa_evaluator.py]: I see. qa_evaluator.py can be combined with yours. I will do it.

File: nce/NCE_MP_Pytorch/model.py (hunk 7 above)
Comment (Victor0118) [reply nce/NCE_MP_Pytorch/model.py]: Sure. Thanks

File: nce/NCE_MP_Pytorch/trainers/trecqa_trainer.py (hunk 10 above)
Comment (Victor0118) [reply nce/NCE_MP_Pytorch/trainers/trecqa_trainer.py]: Great thanks for your suggestion!!!

File: nce/NCE_MP_Pytorch/model.py (hunk 7 above)
Comment (tuzhucheng) [reply nce/NCE_MP_Pytorch/model.py]: Feel free to make changes to MP-CNN model to make the OOP more organized if need be.

**Comment by Victor0118** [comment]
@tuzhucheng  Could you review the code according to the last commit?

File: nce/nce_pairwise_mp/trainers/qa_trainer.py [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+                        for near_id in near_list:
+                            batch_qid.append(qid_i)
+                            batch_aid.append(aid_i)
+
+                            new_train_pos["answer"].append(answer_i)
+                            new_train_pos["question"].append(question_i)
+                            new_train_pos["ext_feat"].append(ext_feat_i)
+
+                            near_answer = self.question2answer[qid_i]["neg"][near_id]["answer"]
+                            if question_i.size()[0] > max_len_q:
```
Comment (tuzhucheng) [inline nce/nce_pairwise_mp/trainers/qa_trainer.py]: can do `.size(0)` to get size for 1st dimension.

File: nce/nce_pairwise_mp/trainers/qa_trainer.py [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+                        # Reference: https://github.com/pytorch/pytorch/issues/264
+                        # loss *= torch.autograd.variable.Variable(torch.from_numpy(np.array(new_near_score)).cuda())
+                        loss *= np.mean(new_near_score)
+
+                    loss_num = loss.data.cpu().numpy()[0]
+                    total_loss += loss_num
+                    loss.backward()
+                    self.optimizer.step()
+
+                    del output
```
Comment (tuzhucheng) [inline nce/nce_pairwise_mp/trainers/qa_trainer.py]: does this help with forcing Python to release memory?

File: utils/nce_neighbors.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+        neg_list.append(key)
+
+    k = min(k, len(neg_dict))
+    min_list = heapq.nsmallest(k, enumerate(dis_list), key=operator.itemgetter(1))
+    min_id_list = [neg_list[x[0]] for x in min_list]
+    return min_id_list
+
+# get the negative samples randomly
+def get_random_neg_id(q2neg, qid_i, k=5):
+    # question 1734 has no neg answer
```
Comment (tuzhucheng) [inline utils/nce_neighbors.py]: which dataset is the comment referring to?

File: utils/nce_neighbors.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    dis_list = []
+    pos_feature = pos_feature.data.cpu().numpy()
+    pos_feature_norm = pos_feature / np.sqrt(sum(pos_feature ** 2))
+    neg_list = []
+    for key in neg_dict:
+        if distance == "l2":
+            dis = np.sqrt(np.sum((np.array(pos_feature) - neg_dict[key]["feature"]) ** 2))
+        elif distance == "cosine":
+            neg_feature = np.array(neg_dict[key]["feature"])
+            feat_norm = neg_feature / np.sqrt(sum(neg_feature ** 2))
```
Comment (tuzhucheng) [inline utils/nce_neighbors.py]: Use `numpy.linalg.norm`? https://stackoverflow.com/a/1401828/2514889

File: nce/nce_pairwise_mp/trainers/qa_trainer.py (hunk 12 above)
Comment (Victor0118) [reply nce/nce_pairwise_mp/trainers/qa_trainer.py]: No in fact. In my case adding it does not solve the Out Of Memory exception but does not affect the result. I should remove it. THX!

File: utils/nce_neighbors.py (hunk 14 above)
Comment (Victor0118) [reply utils/nce_neighbors.py]: I seldom use `numpy.linalg.norm` or `scipy.linalg.norm` because there is a type issue: https://stackoverflow.com/questions/31685156/normalizing-vector-produces-nan-in-numpy
If I write it myself, I do not worry about the possible mistake, right?

File: utils/nce_neighbors.py (hunk 13 above)
Comment (Victor0118) [reply utils/nce_neighbors.py]: TrecQA. I will comment it.

File: datasets/castor_dataset.py [hunk 15]
```diff
@@ -49,3 +50,21 @@ def __init__(self, path, load_ext_feats=False):
...
+            for i, token in enumerate(field.vocab.itos):
+                wv_index = stoi.get(token, None)
+                if wv_index is not None:
+                    field.vocab.vectors[i] = vectors[wv_index]
+                else:
+                    # initialize <unk> with uniform_(-0.05, 0.05) vectors
+                    field.vocab.vectors[i] = torch.FloatTensor(dim).uniform_(-0.05, 0.05)
+        else:
+            print("Error: Need word embedding pt file")
+            exit(1)
```
Comment (tuzhucheng) [inline datasets/castor_dataset.py]: You need to `from sys import exit` before you can do this I think.

File: nce/nce_pairwise_mp/trainers/qa_trainer.py [hunk 16]
```diff
@@ -171,19 +171,18 @@ def train_epoch(self, epoch):
                     loss.backward()
                     self.optimizer.step()
 
-                    del output
-                    del loss
                     del new_train_neg
```
Comment (tuzhucheng) [inline nce/nce_pairwise_mp/trainers/qa_trainer.py]: Deleting `new_train_neg` and `new_train_pos` will have an impact?

File: datasets/castor_dataset.py (hunk 15 above)
Comment (tuzhucheng) [reply datasets/castor_dataset.py]: This is not imported in this file.

Comment (Victor0118) [reply datasets/castor_dataset.py]: https://stackoverflow.com/questions/28621112/if-i-import-exit-from-sys-which-exit-will-it-run
They are different but sys seems safer.

File: nce/nce_pairwise_mp/trainers/qa_trainer.py (hunk 16 above)
Comment (Victor0118) [reply nce/nce_pairwise_mp/trainers/qa_trainer.py]: They contain my self-defined objects with GPU memory. Using del might be safer but I am not clear how it works in the low level environment. 
https://discuss.pytorch.org/t/tensor-to-variable-and-memory-freeing-best-practices/6000/2
